### PR TITLE
PR: Updates underlying imagemeta/meta types

### PR DESCRIPTION
**Pull request by evanoberholster**
This PR updates underlying imagemeta/meta types. 

Breaking changes:
- imagemeta/meta changed ShutterSpeed from [2]uint16 to [2]uint32. 
- imagemeta/jpeg changed ScanJPEG function for improved reader efficiency
- imagemeta/tiff changed Scan function to include imagetype 

Non-breaking changes:
- imagemeta/imagetype added filename extension recognition and improved stringer performance
- imagemeta/meta improved stringer performance 
- imagemeta/meta/canon improved stringer performance (thanks to @abrander)

### CODE REVIEW

File: meta/canon/canon.go [hunk 1]
```diff
@@ -3,47 +3,78 @@ package canon
 
 //go:generate msgp
 
+var (
+	strContinuousDriveDist = []uint{0, 6, 16, 21, 47, 62, 78, 91, 98, 105, 119, 137}
+	strCanonFocusModeDist  = []uint{0, 11, 22, 33, 45, 51, 61, 73}
```
Comment (abrander) [inline meta/canon/canon.go:8]: Would it be possible to use const's for these values?

File: meta/exifTypes.go [hunk 2]
```diff
@@ -1,19 +1,32 @@
...
 
 //go:generate msgp
 
-// FocalLength is a Focal Length expressed in millimeters.
-type FocalLength float32
-
 var (
+	_ExposureModeIndex    = [...]uint8{0, 4, 10, 22}
+	_ExposureProgramIndex = [...]uint8{0, 11, 17, 27, 47, 72, 93, 112, 120, 129, 133}
+	_MeteringModeIndex    = [...]uint8{0, 7, 14, 37, 41, 51, 64, 71}
```
Comment (abrander) [inline meta/exifTypes.go:13]: Maybe add these as consts's too?

**Review by abrander — COMMENTED** [review]
Maybe it's not my place to review this, but I figured that you might want some feedback since you created a PR.

I have only (quickly) reviewed the code and not actually tested the branch.

Feel free to ignore everything :-)

File: meta/canon/canon.go (hunk 1 above)
Comment (evanoberholster) [reply meta/canon/canon.go:8]: Yes. Definitely, Thank you.

File: meta/exifTypes.go (hunk 2 above)
Comment (evanoberholster) [reply meta/exifTypes.go:13]: Yes. I did some benchmarking and this method is used by [https://pkg.go.dev/golang.org/x/tools/cmd/stringer](https://pkg.go.dev/golang.org/x/tools/cmd/stringer) and is significantly more performant than using a map. 

Thanks for the updates.

Comment (abrander) [reply meta/exifTypes.go:13]: My suggestion was regarding the magic value `0, 4, ...` and to on... I would like to replace those by:

```go
const (
	exposureModeUnknown = 0
	exposureModeNormal  = 4
)

var (
	_ExposureModeIndex    = [...]uint8{
		exposureModeUnknown,
		exposureModeNormal,
		...,
	}
)
```
It's much more readable, have no performance-impact and less prone to errors - but it's really not a big deal.

Comment (evanoberholster) [reply meta/exifTypes.go]: I appreciate your suggestions and have made the changes to ExposureMode, ProgramMode, and MeteringMode.

Comment (evanoberholster) [reply meta/exifTypes.go]: The "performance" of looking up a string from 2 slices is faster than a map. this doesn't make a difference unless you scale it to very large numbers. An example of this is serving thousands of exif responses over json.

**Pull request merged**